### PR TITLE
Refactor LLourney to be more self-contained.

### PR DESCRIPTION
**Pull request by tmmnmmn**
Refactor the `LLourney` class to be configured through arguments to `__init__` rather than constants hardcoded in `model.py`.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -208,14 +249,25 @@ def decode_image_latents(self, denoised_latent_image: Tensor) -> np.ndarray:
...
-def forward_pass_test():
-    img = torch.randn([BATCH_SIZE, 3, IMG_DIM, IMG_DIM], device=DEVICE)
-    input_str = """
-    Once upon a time, in the heart of an ancient kingdom, there was a stunning garden filled with vibrant flowers and chirping birds. Amidst the garden, a glistening pond served as a mirror, reflecting the blossoming surroundings. 
-    A gentle breeze rustled the leaves of the towering trees, adding a melodious charm to the scene. The garden was indeed a paradise, a spectacle of nature's magnificence, where every creature lived in har
-    """
-    denoised_latent_img = model(img, [input_str]*BATCH_SIZE, [1]*BATCH_SIZE)
-    # print(denoised_latent_img.shape)
+@torch.no_grad()
+def forward_pass_test(model: LLourney, input_str: str, batch_size: int = 2, device="cpu"):
```
Comment (veezbo) [inline model.py:228]: thanks for getting this working again

File: model.py [hunk 2]
```diff
@@ -224,6 +276,28 @@ def forward_pass_test():
...
+    #     block_out_channels=[32, 64],
+    #     in_channels=3,
+    #     out_channels=3,
+    #     down_block_types=["DownEncoderBlock2D", "DownEncoderBlock2D"],
+    #     up_block_types=["UpDecoderBlock2D", "UpDecoderBlock2D"],
+    #     latent_channels=4,
+    # )
+    test_model = LLourney(
+        image_dim=64,
+        # vae_model_id_or_path="test_vae",
```
Comment (veezbo) [inline model.py]: Do we still need the "test_vae" you defined above around for the test? Or is it the default 'runwayml/stable-diffusion-v1-5' small enough?

File: model.py [hunk 3]
```diff
@@ -208,14 +249,25 @@ def decode_image_latents(self, denoised_latent_image: Tensor) -> np.ndarray:
...
-    input_str = """
-    Once upon a time, in the heart of an ancient kingdom, there was a stunning garden filled with vibrant flowers and chirping birds. Amidst the garden, a glistening pond served as a mirror, reflecting the blossoming surroundings. 
-    A gentle breeze rustled the leaves of the towering trees, adding a melodious charm to the scene. The garden was indeed a paradise, a spectacle of nature's magnificence, where every creature lived in har
-    """
-    denoised_latent_img = model(img, [input_str]*BATCH_SIZE, [1]*BATCH_SIZE)
-    # print(denoised_latent_img.shape)
+@torch.no_grad()
+def forward_pass_test(model: LLourney, input_str: str, batch_size: int = 2, device="cpu"):
+    # Prepare random image in pixel space
+    img = torch.randn((batch_size, 3, model.image_dim, model.image_dim), device=device)
```
Comment (veezbo) [inline model.py:230]: to make this a FloatTensor, could add `.double()` here to make the types line up

File: model.py [hunk 4]
```diff
@@ -1,21 +1,18 @@
...
-from transformers import LlamaTokenizerFast
-from transformers import LlamaModel
 from transformers import AutoTokenizer, AutoModel
 
-from datasets import load_dataset
-
-from typing import Optional
 
 # Input image dimension
 IMG_DIM = 128
```
Comment (veezbo) [inline model.py]: Could you delete the constants no longer needed here?

File: model.py [hunk 5]
```diff
@@ -43,12 +40,38 @@
...
+        llm_tokenizer_model_id_or_path: str = 'gpt2',
+    ):
         super().__init__()
+        self.image_dim = image_dim
+        self.patch_dim = patch_dim
+        self.context_size = context_size
+
+        self.vae = AutoencoderKL.from_pretrained(vae_model_id_or_path, subfolder='vae')
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        self.vae_latent_channels = self.vae.config.latent_channels
```
Comment (veezbo) [inline model.py:37]: Not major, but I might prefer calling this latent_channels or latent_img_channels. Doesn't seem intrinsic to the VAE.

File: model.py [hunk 6]
```diff
@@ -154,41 +195,41 @@ def forward(self, latent_image: Tensor, input_ids: Tensor, timestep: Tensor, tex
...
         
         # Pure reshaping to original image shape, no learnable parameters
         # (B, NP, LC*L*L/NP) -> (B, sqrt[NP], sqrt[NP], L/sqrt[NP], L/sqrt[NP], LC) | B*LC*L^2 on both sides
         # NOTE: NP = (LATENT_DIM // PATCH_DIM)^2
-        projected_img_emb = projected_img_emb.reshape(
-            shape=(B, LATENT_DIM // PATCH_DIM, LATENT_DIM // PATCH_DIM, PATCH_DIM, PATCH_DIM, LATENT_CHANNELS)
+        # projected_img_emb = projected_img_emb.reshape(
+        #     shape=(B, self.latent_dim // self.patch_dim, self.latent_dim // self.patch_dim, self.patch_dim, self.patch_dim, self.vae_latent_channels)
+        # )
+        denoised_latent_image = rearrange(
```
Comment (veezbo) [inline model.py:182]: Can you adjust the shaping comments accordingly? Perhaps explain a bit more about what's going on here, though it's likely a lot clearer than before.

File: model.py [hunk 7]
```diff
@@ -43,12 +40,38 @@
...
+        self.num_patches = (self.latent_dim // self.patch_dim) ** 2
+
+        # TODO(2): Use official tokenizer and model / upgrade the tokenizer and model
+        self.llama_tokenizer = AutoTokenizer.from_pretrained(llm_tokenizer_model_id_or_path)
+        # If the tokenizer doesn't have a pad_token, set it to eos_token
+        if self.llama_tokenizer.pad_token is None:
+            print(f"Setting tokenizer pad_token to eos_token: {self.llama_tokenizer.eos_token}")
+            self.llama_tokenizer.pad_token = self.llama_tokenizer.eos_token
+        # Set padding_side explicitly (?)
+        self.llama_tokenizer.padding_side = "right"
```
Comment (veezbo) [inline model.py:50]: can you leave a comment here suggesting we probably want to try `padding_side = "left"` at some point?

File: model.py [hunk 8]
```diff
@@ -60,41 +83,52 @@ def __init__(self):
...
+        scale_latents: bool = False,
+        generator: Optional[torch.Generator] = None
+    ) -> torch.FloatTensor:
         if self.train:
-            latent_image = self.vae.config.scaling_factor * self.vae.encode(img).latent_dist.sample()
+            latent_image = self.vae.encode(img).latent_dist.sample(generator=generator)
         else:
             latent_image = self.vae.encode(img).latent_dist.mean
+        if scale_latents:
+            latent_image = latent_image * self.vae.config.scaling_factor
```
Comment (veezbo) [inline model.py]: If I'm not mistaken, it would be better to use `self.vae_scale_factor` instead of `self.vae.config.scaling_factor`

File: model.py [hunk 9]
```diff
@@ -60,41 +83,52 @@ def __init__(self):
...
         return latent_image
     
-    @torch.no_grad()
-    def decode_image_latents(self, denoised_latent_image: Tensor, as_numpy: bool = True) -> np.ndarray:
+    def decode_image_latents(self, denoised_latent_image: torch.FloatTensor, as_numpy: bool = True) -> np.ndarray:
         denoised_latent_image = 1 / self.vae.config.scaling_factor * denoised_latent_image
```
Comment (veezbo) [inline model.py]: If I'm not mistaken, it would be better to use `self.vae_scale_factor` instead of `self.vae.config.scaling_factor`

**Review by veezbo — COMMENTED** [review]
Looks good functionally, only a few minor comments

File: model.py [hunk 10]
```diff
@@ -208,14 +249,25 @@ def decode_image_latents(self, denoised_latent_image: Tensor) -> np.ndarray:
...
-    denoised_latent_img = model(img, [input_str]*BATCH_SIZE, [1]*BATCH_SIZE)
-    # print(denoised_latent_img.shape)
+@torch.no_grad()
+def forward_pass_test(model: LLourney, input_str: str, batch_size: int = 2, device="cpu"):
+    # Prepare random image in pixel space
+    img = torch.randn((batch_size, 3, model.image_dim, model.image_dim), device=device)
+
+    # Prepare input tokens and attention mask (for padding)
+    input_str_list = [input_str] * batch_size
+    input_ids, attention_mask = model.tokenize(input_str_list)
```
Comment (veezbo) [inline model.py]: I know it's used as a mask of the attention, but I think calling it `pad_mask` here is more in-line with the output from a tokenizer. Feel free to override though.

File: model.py [hunk 11]
```diff
@@ -60,41 +83,52 @@ def __init__(self):
...
+    def tokenize(self, text, device="cpu", padding="max_length", truncation=True, **tokenizer_kwargs):
+        tokenizer_output = self.llama_tokenizer(
+            text, padding=padding, truncation=truncation, max_length=self.context_size, **tokenizer_kwargs
+        )
+
+        # For now only return input_ids and attention_mask
+        input_ids = torch.tensor(tokenizer_output["input_ids"], device=device, dtype=torch.long)
+        attention_mask = torch.tensor(tokenizer_output["attention_mask"], device=device, dtype=torch.float)
+
+        return input_ids, attention_mask
```
Comment (veezbo) [inline model.py:94]: I know it's used as a mask of the attention, but I think calling it `pad_mask` here is more in-line with the output from a tokenizer. Feel free to override though.

File: model.py [hunk 12]
```diff
@@ -145,7 +186,7 @@ def forward(self, latent_image: Tensor, input_ids: Tensor, timestep: Tensor, tex
         # print("transformer_emb pre llama:", transformer_emb.shape)
 
         # Push unified embedding through Llama transformer blocks
-        total_pad_mask = F.pad(text_pad_mask, (0, NUM_PATCHES+1), value=1)  # (B, T+1+NP)
+        total_pad_mask = F.pad(text_pad_mask, (0, self.num_patches + 1), value=1)  # (B, T+1+NP)
         transformer_emb = self.llama(inputs_embeds=transformer_emb, attention_mask=total_pad_mask).last_hidden_state  # (B, S, C)
         # print("transformer_emb post llama:", transformer_emb.shape)
```
Comment (veezbo) [inline model.py:169]: Can delete the below TODO as you just did it

File: model.py [hunk 13]
```diff
@@ -145,7 +186,7 @@ def forward(self, latent_image: Tensor, input_ids: Tensor, timestep: Tensor, tex
         # print("transformer_emb pre llama:", transformer_emb.shape)
```
Comment (veezbo) [inline model.py:163]: Can delete the above TODOs I believe as you resolved them.

File: model.py (hunk 2 above)
Comment (tmmnmmn) [reply model.py]: I guess the current `LLourney` parameters for the test are meant to work for a small VAE; I believe if you use a 512 x 512 image and the VAE from `runwayml/stable-diffusion-v1-5`, the patch embedding (with `patch_dim=2`) will produce way more tokens than the `hf-internal-testing/tiny-random-gpt2` testing checkpoint's context window can handle. So it's probably best to stick with `test_vae`, and it makes the test faster and less resource-intensive to run since it's smaller. (I guess it could be worth uploading it to the hub so we don't have to keep around the code to reproduce it.)

File: model.py (hunk 3 above)
Comment (tmmnmmn) [reply model.py:230]: I'm not sure if I understand the comment, [`torch.randn`](https://pytorch.org/docs/stable/generated/torch.randn.html) should use `dtype = torch.float (= torch.float32)` by default. So it's already a `torch.FloatTensor`, and I'm not sure casting it to a [double](https://pytorch.org/docs/stable/generated/torch.Tensor.double.html) (= `torch.float64`) would make sense here.

File: model.py (hunk 8 above)
Comment (tmmnmmn) [reply model.py]: `self.vae_scale_factor` and `self.vae.config.scaling_factor` are two different quantities: `self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1) (= 8)` is the factor by which the VAE downsamples the spatial resolution of the input image, whereas [`self.vae.config.scaling_factor (= 0.1825)`](https://huggingface.co/docs/diffusers/api/models/autoencoderkl#diffusers.AutoencoderKL.scaling_factor) is a scalar which scales the output of the VAE encoder. So I believe using `self.vae.config.scaling_factor` is correct here.

File: model.py (hunk 9 above)
Comment (tmmnmmn) [reply model.py]: See https://github.com/veezbo/llourney-dev/pull/1#discussion_r1307264135. I believe using `self.vae.config.scaling_factor` is correct here.

File: model.py (hunk 3 above)
Comment (veezbo) [reply model.py:230]: Basically Pycharm complains at the typing (it doesn't believe `img` is a FloatTensor) which could be its own fault. But adding `.double()` suppresses that error. Not a big deal in any case.

File: model.py (hunk 8 above)
Comment (veezbo) [reply model.py]: I didn't realize `self.vae_scale_factor` and `self.vae.config.scaling_factor` were different. Can you rename them so their respective usages are more clear?

File: model.py (hunk 2 above)
Comment (veezbo) [reply model.py]: Sounds good to use it for the test, but then should we uncomment out the code that creates/saves it? Maybe with a check to see whether it has already been saved.

**Review by veezbo — APPROVED** [review]
one renaming request and maybe a couple minor suggestions. looks good

File: model.py (hunk 11 above)
Comment (tmmnmmn) [reply model.py:94]: I think I might prefer `attention_mask` here because in general a LLM tokenizer might produce a more complicated attention mask.

File: model.py (hunk 2 above)
Comment (tmmnmmn) [reply model.py]: Added a simple check to see if the `test_vae` directory exists and if not generate the random small VAE checkpoint for testing.

**Review by veezbo — APPROVED** [review]

**Comment by veezbo** [comment]
@dg845 feel free to merge when you're ready
